jena-arq/src/main/java/org/apache/jena/riot/system/AsyncParser.java

```diff
@@ -239,40 +278,143 @@ public void fatal(String message, long line, long col) {
                 throw new RiotException(SysRIOT.fmtMessage(message, line, col)) ;
             }
         };
+    }
+
+    /** Returns a runnable for stopping the parse process */
+    static Runnable startParserThread(
+            Logger LOG1, List<RDFParserBuilder> parserBuilders,
+            BlockingQueue<List<EltStreamRDFImpl>> queue,
+            int chunkSize,
+            boolean daemonMode) {
+
+        // A flag for whether parsing is considered aborted - regardless
+        // whether due to normal or exceptional conditions
+        AtomicBoolean isAborted = new AtomicBoolean(false);
+
+        // -- Parser thread setup
+        Consumer<List<EltStreamRDFImpl>> destination = batch -> {
+            if (!isAborted.get()) {
+                try {
+                    queue.put(batch);
+                } catch (InterruptedException ex) {
+                    // FmtLog.error(LOG, ex, "Error: %s", ex.getMessage());
+                    throw new RuntimeException(ex);
+                }
+            } else {
+                throw new RuntimeException(new InterruptedException());
+            }
+        };
+        EltStreamBatcher batcher = new EltStreamBatcher(destination, chunkSize);
+        StreamRDF generatorStream = new StreamToElements(batcher);
 
         // Parser thread
-        Runnable task = ()->{
-            batcher.startBatching();
+        Runnable task = () -> {
+            batcher.startBatching(); // No try/catch block currently needed because this is a no-op
```

Comment:
startBatching and finishBatching no longer line up.
finishBatching is conditional. Can it ever fail to be called? Comments needed.

Reply:
If close() is invoked and isAborted is false then isAborted is atomically set to true. Furthermore, the close()-logic proceeds to clear the queue and place the poison on it.
I essence, if isAborted is true, then it means that the poison is (about to be) placed on the queue - and therefore finishBatching is no longer needed.

In that case finishBatching is suppressed because it would redundantly try to place another poison on the queue.

Reply:
But does it do any harm?

Thinking long term - clear code such as "start-finish" pairing, or "start-finish|abort" or "start-shutdown(current "finish")-finish" is useful to have for the next person (who may be yourself in a few PRs time!)

Reply:
> But does it do any harm?

Right now the chunk-destination raises an exception if it is in abort state intended to forcefully kill the parser thread once it sends out another chunk.
Calling finishBatching with isAborted=true would attempt to dispatch the remaining batch and poison to the destination which is guaranteed to raise an exception.
Maybe in order to improve the structure, the `Runnable task = ()` would have to be factored out into its own class (that implements Runnable), so that parts of the process can be put into their own methods.

Reply:
I factored out `Runnable task = () -> {...};` into its own `Task` class. There the run method has a clearer structure making use of private helper methods.